src/global/install.rs

```diff
@@ -444,63 +503,122 @@ pub(crate) async fn sync(
-            .map(|channel| channel.clone().into_channel(config.global_channel_config()))
-            .collect_vec();
-
-        let repodata = await_in_progress("querying repodata ", |_| async {
-            gateway
-                .query(
-                    channels,
-                    [environment.platform(), Platform::NoArch],
-                    specs.values().cloned().collect_vec(),
-                )
-                .recursive(true)
-                .await
-                .into_diagnostic()
-        })
-        .await?;
+        let env_dir = EnvDir::new(env_root.clone(), env_name.clone()).await?;
+        let prefix = Prefix::new(env_dir.path());
 
-        // Determine virtual packages of the current platform
-        let virtual_packages = VirtualPackage::detect(&VirtualPackageOverrides::default())
-            .into_diagnostic()
-            .context("failed to determine virtual packages")?
-            .iter()
-            .cloned()
-            .map(GenericVirtualPackage::from)
-            .collect();
-
-        // Solve the environment
-        let solver_specs = specs.clone();
-        let solved_records = tokio::task::spawn_blocking(move || {
-            wrap_in_progress("solving environment", move || {
-                Solver.solve(SolverTask {
-                    specs: solver_specs.values().cloned().collect_vec(),
-                    virtual_packages,
-                    ..SolverTask::from_iter(&repodata)
-                })
-            })
-            .into_diagnostic()
-            .context("failed to solve environment")
-        })
-        .await
-        .into_diagnostic()??;
-
-        let packages = specs.keys().cloned().collect();
-
-        install_environment(
-            &environment_name,
-            &environment.exposed,
-            packages,
-            solved_records.clone(),
-            auth_client.clone(),
-            environment.platform(),
-            bin_dir,
+        let prefix_records = prefix.find_installed_packages(Some(50)).await?;
+
+        if !specs_match_local_environment(&specs, prefix_records, parsed_environment.platform()) {
+            install_environment(
+                &specs,
+                &env_name,
+                &parsed_environment,
+                auth_client.clone(),
+                &prefix,
+                config,
+                &gateway,
+            )
+            .await?;
+        }
+
+        expose_executables(
+            &env_name,
+            &parsed_environment,
+            specs.keys().cloned().collect(),
+            &prefix,
+            &bin_dir,
         )
         .await?;
     }
 
     Ok(())
 }
+
+/// Checks if the local environment matches the given specifications.
+///
+/// This function verifies that all the given specifications are present in the
+/// local environment's prefix records and that there are no extra entries in
+/// the prefix records that do not match any of the specifications.
+fn specs_match_local_environment(
```

Comment:
Ideally you would write some tests to verify this specific function.

You can use a lock-file to extract a set of `RepoDataRecord`s. If you make this function generic over `AsRef<RepoDataRecord>` you can use both `PrefixRecord` and plain `RepoDataRecord`s.